tests/unit/s2n_pq_kem_test.c

```diff
@@ -0,0 +1,119 @@
+                .asm_is_enabled = s2n_pq_no_asm_available,
+                .enable_asm = s2n_pq_noop_asm,
+                .disable_asm = s2n_pq_noop_asm,
+        },
+        {
+                .kem = &s2n_sike_p434_r2,
+                .asm_is_enabled = s2n_sikep434r2_asm_is_enabled,
+                .enable_asm = s2n_try_enable_sikep434r2_asm,
+                .disable_asm = s2n_disable_sikep434r2_asm,
+        },
+        {
+                .kem = &s2n_kyber_512_r2,
+                .asm_is_enabled = s2n_pq_no_asm_available,
+                .enable_asm = s2n_pq_noop_asm,
+                .disable_asm = s2n_pq_noop_asm,
+        },
+        {
+                .kem = &s2n_kyber_512_90s_r2,
+                .asm_is_enabled = s2n_pq_no_asm_available,
+                .enable_asm = s2n_pq_noop_asm,
+                .disable_asm = s2n_pq_noop_asm,
+        },
+};
+
+int main() {
+    BEGIN_TEST();
+
+    for (size_t i = 0; i < s2n_array_len(test_vectors); i++) {
+        const struct s2n_kem_test_vector vector = test_vectors[i];
+        const struct s2n_kem *kem = vector.kem;
+
+        uint8_t *public_key = (uint8_t *)malloc(kem->public_key_length);
+        uint8_t *private_key = (uint8_t *)malloc(kem->private_key_length);
+        uint8_t *client_shared_secret = (uint8_t *)malloc(kem->shared_secret_key_length);
+        uint8_t *server_shared_secret = (uint8_t *)malloc(kem->shared_secret_key_length);
+        uint8_t *ciphertext = (uint8_t *)malloc(kem->ciphertext_length);
+
+        if (s2n_pq_is_enabled()) {
+            /* Test the C code */
+            EXPECT_OK(vector.disable_asm());
+            EXPECT_SUCCESS(kem->generate_keypair(public_key, private_key));
+            EXPECT_SUCCESS(kem->encapsulate(ciphertext, client_shared_secret, public_key));
+            EXPECT_SUCCESS(kem->decapsulate(server_shared_secret, ciphertext, private_key));
+            EXPECT_BYTEARRAY_EQUAL(server_shared_secret, client_shared_secret, kem->shared_secret_key_length);
+
+            /* Test the assembly, if available */
+            EXPECT_OK(vector.enable_asm());
+            if (vector.asm_is_enabled()) {
+                memset_check(public_key, 0, kem->public_key_length);
+                memset_check(private_key, 0, kem->private_key_length);
+                memset_check(client_shared_secret, 0, kem->shared_secret_key_length);
+                memset_check(server_shared_secret, 0, kem->shared_secret_key_length);
+                memset_check(ciphertext, 0, kem->ciphertext_length);
+
+                EXPECT_SUCCESS(kem->generate_keypair(public_key, private_key));
+                EXPECT_SUCCESS(kem->encapsulate(ciphertext, client_shared_secret, public_key));
+                EXPECT_SUCCESS(kem->decapsulate(server_shared_secret, ciphertext, private_key));
+                EXPECT_BYTEARRAY_EQUAL(server_shared_secret, client_shared_secret, kem->shared_secret_key_length);
+            }
+        } else {
+#if defined(S2N_NO_PQ)
```

Comment:
What behavior do you expect if s2n_pq_is_enabled() == false but S2N_NO_PQ is not define?

Reply:
In that case (when I originally wrote the test), I suppose it would have been possible to still call the PQ KEM functions and have them succeed. But, that's not the behavior we want. I added checks to each of the KEM's keygen/enc/dec function to ensure PQ is enabled.

Reply:
So, adding that check in the KEM functions right now breaks other tests when testing in FIPS mode. The plan is to put those checks in and fix all the other tests, but I don't want to inflate this PR with those changes. This PR, as-is, maintains all of our existing test cases (and adds the `S2N_ERR_UNIMPLEMENTED` checks); I'd like to keep this PR scoped to just those change and follow it up with a separate PR to address the other test fixes.

Reply:
Can you open an issue to track that problem? I want to make sure we don't lose it, especially since it's not obvious :)

Reply:
https://github.com/awslabs/s2n/issues/2474